pkg/ansible/jobgenerator.go

```diff
@@ -0,0 +1,191 @@
+
+	playbookPath := path.Join(openshiftAnsibleContainerDir, playbook)
+	env := []kapi.EnvVar{
+		{
+			Name:  "INVENTORY_FILE",
+			Value: "/ansible/inventory/hosts",
+		},
+		{
+			Name:  "PLAYBOOK_FILE",
+			Value: playbookPath,
+		},
+		{
+			Name:  "ANSIBLE_HOST_KEY_CHECKING",
+			Value: "False",
+		},
+		{
+			Name:  "OPTS",
+			Value: "-vvv --private-key=/ansible/ssh/privatekey.pem -e @/ansible/inventory/vars",
+		},
+	}
+
+	if cluster.Spec.Hardware.AWS != nil && len(cluster.Spec.Hardware.AWS.AccountSecret.Name) > 0 {
+		env = append(env, []kapi.EnvVar{
+			{
+				Name: "AWS_ACCESS_KEY_ID",
+				ValueFrom: &kapi.EnvVarSource{
+					SecretKeyRef: &kapi.SecretKeySelector{
+						LocalObjectReference: cluster.Spec.Hardware.AWS.AccountSecret,
+						Key:                  "aws_access_key_id",
+					},
+				},
+			},
+			{
+				Name: "AWS_SECRET_ACCESS_KEY",
+				ValueFrom: &kapi.EnvVarSource{
+					SecretKeyRef: &kapi.SecretKeySelector{
+						LocalObjectReference: cluster.Spec.Hardware.AWS.AccountSecret,
+						Key:                  "aws_secret_access_key",
+					},
+				},
+			},
+		}...)
+	}
+
+	sshKeyFileMode := int32(0600)
+
+	podSpec := kapi.PodSpec{
+		DNSPolicy:     kapi.DNSClusterFirst,
+		RestartPolicy: kapi.RestartPolicyNever,
+
+		Containers: []kapi.Container{
+			{
+				Name:            "ansible",
+				Image:           r.Image,
+				ImagePullPolicy: r.ImagePullPolicy,
+				Env:             env,
+				VolumeMounts: []kapi.VolumeMount{
+					{
+						Name:      "inventory",
+						MountPath: "/ansible/inventory/",
+					},
+				},
+			},
+		},
+		Volumes: []kapi.Volume{
+			{
+				// Mounts both our inventory and vars file.
+				Name: "inventory",
+				VolumeSource: kapi.VolumeSource{
+					ConfigMap: &kapi.ConfigMapVolumeSource{
+						LocalObjectReference: kapi.LocalObjectReference{
+							Name: cfgMap.Name,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	if cluster.Spec.Hardware.AWS != nil && len(cluster.Spec.Hardware.AWS.SSHSecret.Name) > 0 {
+		podSpec.Containers[0].VolumeMounts = append(podSpec.Containers[0].VolumeMounts, kapi.VolumeMount{
+			Name:      "sshkey",
+			MountPath: "/ansible/ssh/",
+		})
+		podSpec.Volumes = append(podSpec.Volumes, kapi.Volume{
+			Name: "sshkey",
+			VolumeSource: kapi.VolumeSource{
+				Secret: &kapi.SecretVolumeSource{
+					SecretName: cluster.Spec.Hardware.AWS.SSHSecret.Name,
+					Items: []kapi.KeyToPath{
+						{
+							Key:  "ssh-privatekey",
+							Path: "privatekey.pem",
+							Mode: &sshKeyFileMode,
+						},
+					},
+				},
+			},
+		})
+	}
+
+	completions := int32(1)
+	deadline := int64(60 * 60) // one hour for now
+
+	job := &kbatch.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            name,
+			OwnerReferences: []metav1.OwnerReference{*clusterOwnerRef},
+		},
+		Spec: kbatch.JobSpec{
+			Completions:           &completions,
+			ActiveDeadlineSeconds: &deadline,
+			Template: kapi.PodTemplateSpec{
+				Spec: podSpec,
+			},
+		},
+	}
+
+	return job, cfgMap
```

Comment:
Returning both rather than creating the config map as well. 👍

Reply:
This now looks extremely unit testable. Should we do this now or perhaps add a trello card to add them in future?

Reply:
I haven't added any unit tests at all for this pr yet, but we can do in a follow up pr if that's ok